src/ipld/util.rs

```diff
@@ -448,102 +455,87 @@ impl<DB: Blockstore + Send + Sync + 'static, T: Borrow<Tipset>, ITER: Iterator<I
-                                // This will simply return an error once we reach that item in
-                                // the queue.
-                            } else if *this.fail_on_dead_links {
-                                this.queue.push(block.state_root);
-                            } else {
-                                // Make sure we update seen here as we don't send the block for
-                                // inspection.
-                                this.seen.lock().insert(block.state_root);
+                                // Visit the block if it's within required depth. And a special case for `0`
+                                // epoch to match Lotus' implementation.
+                                if (block.epoch == 0 || block.epoch > stateroot_limit)
+                                    && should_save_block_to_snapshot(block.state_root)
+                                {
+                                    if this.db.has(&block.state_root)? {
+                                        this.extract_sender.send(block.state_root)?;
+                                        // This will simply return an error once we reach that item in
+                                        // the queue.
+                                    } else if fail_on_dead_links {
+                                        this.queue.push((block.state_root, None));
+                                    } else {
+                                        // Make sure we update seen here as we don't send the block for
+                                        // inspection.
+                                        this.seen.lock().insert(block.state_root);
+                                    }
+                                }
                             }
                         }
-                    }
-                }
-            } else {
-                match this.block_receiver.try_recv() {
-                    Ok(item) => return Poll::Ready(Some(item)),
-                    Err(err) => {
-                        if this.extract_sender.is_empty() {
-                            this.worker_handle.abort();
-                            return Poll::Ready(None);
-                            // This should never happen, because both `extract_sender` and
-                            // `block_receiver` are held by worker_handle and their counterparts -
-                            // by the main process. So those are either both functional or both
-                            // closed.
-                        } else if err == TryRecvError::Disconnected {
-                            panic!(
-                                "block_receiver can only be closed after extract_sender is empty"
-                            )
-                        }
+                    } else if this.extract_sender.is_empty() {
+                        this.worker_handle.abort();
```

Comment:
Right now we are aborting without checking if the `workers` are still processing the existing load.

Do you think we should make sure that all the workers have finished before aborting it here, we can send signal from the worker side (once all tipset tasks are done) and then abort.

Reply:
Great point. The old logic did not handle the shutdown gracefully indeed. Fixed by closing the `extract_sender` instead of forcing `worker_handle.abort`